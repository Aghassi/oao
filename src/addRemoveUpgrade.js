// @flow

import { merge, set as timmSet } from 'timm';
import { mainStory } from 'storyboard';
import kebabCase from 'kebab-case';
import { readAllSpecs, readOneSpec } from './utils/readSpecs';
import removeInternalLinks from './utils/removeInternalLinks';
import writeSpecs from './utils/writeSpecs';
import { exec } from './utils/shell';

const PASS_THROUGH_OPTS = ['dev', 'peer', 'optional', 'exact', 'tilde', 'ignoreEngines'];

type Operation = 'add' | 'remove' | 'upgrade';
type Options = {
  src: string,
  dev?: boolean,
  peer?: boolean,
  optional?: boolean,
  exact?: boolean,
  tilde?: boolean,
  ignoreEngines?: boolean,
};

const run = async (pkgName: string, op: Operation, deps: Array<string>, opts: Options) => {
  const { src: srcPatterns } = opts;
  const allSpecs = await readAllSpecs(srcPatterns);
  if (!allSpecs[pkgName]) {
    mainStory.error(`No such package: ${pkgName}`);
    process.exit(1);
  }

  const pkgNames = Object.keys(allSpecs);
  const { pkgPath, specPath, specs: prevSpecs } = allSpecs[pkgName];
  let succeeded = false;
  const { nextSpecs, removedPackagesByType } = removeInternalLinks(prevSpecs, pkgNames);
  try {
    if (nextSpecs !== prevSpecs) writeSpecs(specPath, nextSpecs);
    mainStory.info(`Executing 'yarn ${op}'...`);
    let cmd = `yarn ${op}`;
    if (deps.length) cmd += ` ${deps.join(' ')}`;
    PASS_THROUGH_OPTS.forEach((key) => { if (opts[key]) cmd += ` --${kebabCase(key)}`; });
    await exec(cmd, { cwd: pkgPath });
    succeeded = true;
  } catch (err) { /* ignore */ }

  // If not successful, revert to the original specs
  if (!succeeded) {
    if (prevSpecs != null) writeSpecs(specPath, prevSpecs);
    return;
  }

  // Read the updated package.json, and add the internal deps
  const { specs: updatedSpecs } = readOneSpec(pkgPath);
  let finalSpecs = updatedSpecs;
  Object.keys(removedPackagesByType).forEach((type) => {
    const removedPackages = removedPackagesByType[type];
    const nextDeps = merge((updatedSpecs[type] || {}), removedPackages);
    finalSpecs = timmSet(finalSpecs, type, nextDeps);
  });
  writeSpecs(specPath, finalSpecs);
};

export default run;
