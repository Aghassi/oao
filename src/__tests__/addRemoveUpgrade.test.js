/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import addRemoveUpgrade from '../addRemoveUpgrade';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

describe('ADD/REMOVE/UPGRADE commands', () => {
  it('touches only the correct package.json', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await addRemoveUpgrade('oao-b', 'add', ['mady'], { src: 'test/fixtures/packages2/*' });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();  // removes internal deps
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();  // restores internal deps
  });

  it('touches only the correct package.json, with custom links', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await addRemoveUpgrade('oao-b', 'add', ['mady'], {
      src: 'test/fixtures/packagesCustomLinks/*',
      link: 'ext-.*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();  // removes internal deps
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();  // restores internal deps
  });

  it('executes ADD correctly (one package, no flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'add', ['mady'], { src: 'test/fixtures/packages2/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes ADD correctly (multiple packages, no flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'add', ['mady', 'jest-html'], { src: 'test/fixtures/packages2/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes ADD correctly (flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'add', ['mady', 'jest-html'], {
      src: 'test/fixtures/packages2/*',
      dev: true,
      exact: true,
    });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes REMOVE correctly', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'remove', ['mady'], { src: 'test/fixtures/packages2/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes UPGRADE correctly (one package, no flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'upgrade', ['mady'], { src: 'test/fixtures/packages2/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes UPGRADE correctly (one package, flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'upgrade', ['mady'], {
      src: 'test/fixtures/packages2/*',
      ignoreEngines: true,
    });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes UPGRADE correctly (no package, no flags)', async () => {
    const helpers = require('../utils/shell');
    await addRemoveUpgrade('oao-b', 'upgrade', [], { src: 'test/fixtures/packages2/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
