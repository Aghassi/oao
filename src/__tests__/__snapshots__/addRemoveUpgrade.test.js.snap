exports[`ADD/REMOVE/UPGRADE commands executes ADD correctly (flags) 1`] = `
Array [
  Array [
    "yarn add mady jest-html --dev --exact",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes ADD correctly (multiple packages, no flags) 1`] = `
Array [
  Array [
    "yarn add mady jest-html",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes ADD correctly (one package, no flags) 1`] = `
Array [
  Array [
    "yarn add mady",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes REMOVE correctly 1`] = `
Array [
  Array [
    "yarn remove mady",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes UPGRADE correctly (no package, no flags) 1`] = `
Array [
  Array [
    "yarn upgrade",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes UPGRADE correctly (one package, flags) 1`] = `
Array [
  Array [
    "yarn upgrade mady --ignore-engines",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands executes UPGRADE correctly (one package, no flags) 1`] = `
Array [
  Array [
    "yarn upgrade mady",
    Object {
      "cwd": "test/fixtures/packages2/oao-b",
    },
  ],
]
`;

exports[`ADD/REMOVE/UPGRADE commands touches only the correct package.json 1`] = `
Object {
  "dependencies": Object {
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;

exports[`ADD/REMOVE/UPGRADE commands touches only the correct package.json 2`] = `
Object {
  "dependencies": Object {
    "oao": "*",
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;

exports[`ADD/REMOVE/UPGRADE commands touches only the correct package.json, with custom links 1`] = `
Object {
  "dependencies": Object {
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;

exports[`ADD/REMOVE/UPGRADE commands touches only the correct package.json, with custom links 2`] = `
Object {
  "dependencies": Object {
    "ext-one": "*",
    "oao": "*",
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;
